### PR TITLE
Speed seeding for HLT

### PR DESCRIPTION
**Pull request by VinInn**
Speeding up seeding in LHT for electron and regional
negligible regressions expected due to numerics

fyi @mtosi @fwyzard @matteosan1 

offline mtv at
http://innocent.home.cern.ch/innocent/RelVal/pu35_81_fhlt

candidate to back port if hlt asks for

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

File: DataFormats/GeometryVector/interface/Pi.h [hunk 1]
```diff
@@ -28,13 +28,13 @@
 
 namespace Geom {
 
-  inline double pi()     {return     3.141592653589793238;}
-  inline double twoPi()  {return 2. *3.141592653589793238;}
-  inline double halfPi() {return 0.5*3.141592653589793238;}
+  inline constexpr double pi()     {return     3.141592653589793238;}
+  inline constexpr double twoPi()  {return 2. *3.141592653589793238;}
+  inline constexpr double halfPi() {return 0.5*3.141592653589793238;}
```
Comment (fwyzard) [inline DataFormats/GeometryVector/interface/Pi.h]: why not use M_PI, 2. * M_PI, M_PI_2 ?

Comment (VinInn) [inline DataFormats/GeometryVector/interface/Pi.h]: because is like this since orca times...
(somewhere else you will find CLHEP::pi())
I would not mind to go back to old good C macros
(as already done elsewhere)

File: RecoEgamma/EgammaElectronAlgos/interface/ElectronUtilities.h [hunk 2]
```diff
@@ -34,10 +34,10 @@ class ExceptionSafeStlPtrCol : public StlColType
 template <typename RealType>
 RealType normalized_phi( RealType phi )
  {
-  if (phi>CLHEP::pi)
-   { phi -= (2*CLHEP::pi) ; }
-  if (phi<-CLHEP::pi)
-   { phi += (2*CLHEP::pi) ; }
+  constexpr RealType pi(M_PI);
```
Comment (VinInn) [inline RecoEgamma/EgammaElectronAlgos/interface/ElectronUtilities.h]: @fwyzard lke here.

**Comment by VinInn** [comment]
cannot find the failed test
is it not a das issue?

**Comment by VinInn** [comment]
comparison shows no regression

**Comment by VinInn** [comment]
I noticed the truntestTqafTopEventSelection fails also in  #13820

File: Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc [hunk 3]
```diff
@@ -129,6 +129,8 @@ void TrackerGeometry::addDetUnitId(DetId p){
 }
 
 void TrackerGeometry::addDet(GeomDet const * p) {
+  // set index
+  const_cast<GeomDet *>(p)->setGdetIndex(theDets.size());
```
Comment (slava77) [inline Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc]: static analyzer is not happy with (yet another) const_cast in this file.
Why are they even here, should the method signature be changed instead?

File: RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc [hunk 4]
```diff
@@ -47,6 +47,19 @@
...
 #include <utility>
 
+namespace {
+  struct Count {
+    long long s=0;
+    long long n=0;
+    ~Count() { std::cout << "ElectronSeeds res " << s<<'/'<<n << std::endl;}
+  };
+
+  Count stcount;
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc]: static analyzer is not happy with  ``` non-const global static variable ```

Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc]: log files come out with e.g. ``` ElectronSeeds res 39/206 ```.
This should be wrapped in some ifdef SOMEDEBUG.

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc [hunk 5]
```diff
@@ -12,10 +11,21 @@
...
 using namespace reco ;
 using namespace std ;
 
+namespace {
+  struct Count { 
+    long long n=0;
+    ~Count() { std::cout << "PixelHitMatcher res " << n << std::endl;}
+  };
+
+  Count stcount;
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: another debug counter to be removed or hidden with debug flags

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc [hunk 6]
```diff
@@ -106,52 +114,79 @@ PixelHitMatcher::compatibleSeeds
    typedef std::unordered_map<std::pair<const GeomDet*,GlobalPoint>, TrajectoryStateOnSurface> PosTsosAssoc;
    const int charge = int(fcharge) ;
 
-   const double xmeas_phi = xmeas.phi();
+   // auto xmeas_phi = xmeas.barePhi();
+   auto xmeas_r = xmeas.perp();
+   
+   const float phicut = std::cos(2.5);
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: does the 2.5 have any specific meaning?
If it's just a reasonable value magic number, why not use "-0.8" ? It will at least not looks like a constant computed every time.

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc [hunk 7]
```diff
@@ -106,52 +114,79 @@ PixelHitMatcher::compatibleSeeds
...
-      const unsigned idx1 = std::distance(hits.first,it1);
+      auto  idx1 = std::distance(hits.first,it1);
       const DetId id1 = it1->geographicalId();
-      const GeomDet *geomdet1 = it1->det();      
+      const GeomDet *geomdet1 = it1->det();
+
+      auto ix1 = geomdet1->gdetIndex();
+      assert(ix1<int(ndets));
+      /*
+      if (away[ix1]) continue;
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: is this needed here?
if needed, please add a comment in the code why

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc [hunk 8]
```diff
@@ -106,52 +114,79 @@ PixelHitMatcher::compatibleSeeds
...
+        vTsos.push_back(prop1stLayer->propagate(tsos,geomdet1->surface()));
       }
+      auto tsos1 = &vTsos[iTsos[ix1]];
+
       if( !tsos1->isValid() ) continue;
       std::pair<bool, double> est = ( id1.subdetId() % 2 ? 
 				      meas1stBLayer.estimate(vprim, *tsos1, hit1Pos) :
 				      meas1stFLayer.estimate(vprim, *tsos1, hit1Pos)  );
       if( !est.first ) continue;
-      if( std::abs(normalized_phi(hit1Pos.phi()-xmeas_phi))>2.5 ) continue;
```
Comment (slava77) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: ah, this is the 2.5 from https://github.com/cms-sw/cmssw/pull/13819/files#r57508991
.. OK then.

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc (hunk 5 above)
Comment (VinInn) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: forgot to remove, will be done

**Comment by VinInn** [comment]
I will address all raised issues when back to work (late) next week together with those coming from the review of #13831

**Comment by slava77** [comment]
Some observations from comparisons, based on 
https://github.com/cms-sw/cmssw/pull/13819	38fa966
in CMSSW_8_1_X_2016-03-23-2300 (timing compared in CMSSW_8_1_X_2016-03-28-2300).
based on higher stat tests.
Changes are small and are limited to 
_electronMergedSeeds\|_gedGsfElectrons\|_electronGsfTracks (decreasing in significance from seeds to the electrons)
with a single-bin difference in photon sigmaEtaEta as well.

Looking at 25202 for technical performance changes: timing is reduced in ecalDrivenElectronSeeds and similar modules.
- step3 (reco)
```
  delta/mean delta/orJob     original                   new       module name
   -1.057062      -0.09%       109.18 ms/ev ->        33.68 ms/ev ecalDrivenElectronSeeds
```
- step2 (HLT)
```
  delta/mean delta/orJob     original                   new       module name
   -1.014807      -0.64%       622.82 ms/ev ->       203.53 ms/ev hltEgammaElectronPixelSeeds
   -0.616037      -0.13%       181.59 ms/ev ->        96.07 ms/ev hltIter2HighPtPixelSeeds
   -0.583892      -0.10%       146.55 ms/ev ->        80.32 ms/ev hltIter1HighPtPixelSeeds
   -0.514827      -0.02%        31.50 ms/ev ->        18.60 ms/ev hltIter2PFlowPixelSeeds
   -0.463526      -0.04%        68.56 ms/ev ->        42.76 ms/ev hltIter1PFlowPixelSeeds
   -0.455599      -0.11%       186.90 ms/ev ->       117.54 ms/ev hltIter3HighPtMixedSeeds
   -0.414885      -0.01%        14.53 ms/ev ->         9.53 ms/ev hltIter1ElectronsPixelSeeds
   -0.208891      -0.01%        32.01 ms/ev ->        25.96 ms/ev hltIter2ElectronsPixelSeeds
   -0.144136      -0.02%       104.58 ms/ev ->        90.52 ms/ev hltParticleFlowRecHitPSUnseeded
   -0.098660      -0.05%       374.15 ms/ev ->       338.97 ms/ev hltIter4HighPtPixelLessSeeds
```

This behavior is as expected.

Now waiting for an update to remove text printouts leftover from debugging steps.

File: Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc (hunk 3 above)
Comment (VinInn) [inline Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc]: it would require modifications in the whole TrackerGeometry building code...
let's wait it stabilize for PhaseII, then we can clean it up.

File: RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc (hunk 7 above)
Comment (VinInn) [inline RecoEgamma/EgammaElectronAlgos/src/PixelHitMatcher.cc]: it generates regression w/r/t 8_0_0	 in my opinion it is safe and makes	sense 
I propose to leave it and address it together with Egamma folks

**Comment by VinInn** [comment]
@cmsbuild , please test

File: Geometry/CommonDetUnit/interface/GeomDet.h [hunk 9]
```diff
@@ -119,9 +125,10 @@ class GeomDet {
 
   ReferenceCountingPointer<Plane>  thePlane;
   DetId m_detId;
-  int m_index;
+  int m_index=-1;
```
Comment (ianna) [inline Geometry/CommonDetUnit/interface/GeomDet.h]: @VinInn - it would be nice to clean up the class code a bit: sort public, protected and private members, name the latter either with leading m_* or the*, indent the code consistently.

Comment (VinInn) [inline Geometry/CommonDetUnit/interface/GeomDet.h]: It would be nice to cleanup all old code in general yes.
takes time and effort.
Let's schedule it for PhaseI (this code is candidate for backport to 800).

**Comment by VinInn** [comment]
is this runtestTqafTopEventSelection failing randomly? last time was in  #13831 ... (which now passed tests with full marks!)

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/13819	426feb3
- changes in the code are in line with the description
- see performance summary as of https://github.com/cms-sw/cmssw/pull/13819	38fa966 in https://github.com/cms-sw/cmssw/pull/13819#issuecomment-204078373
- the latest update is fully in 426feb3 (the rest is a rebase of the old commits) with changes cleaning up counting code used for debugging and reformatting of LogDebug messages
    - log files in steps with PixelHitMatcher and ElectronSeedGenerator do not have the extra printouts at the end anymore.

**Comment by ianna** [comment]
+1

**Comment by VinInn** [comment]
not sure why simulation is involved, @civanch I am sure nothing should change in sim

**Comment by civanch** [comment]
+1
There is no direct touch to Sim, 
Concerning failing of the unit test: many PRs passing standard tests without a problem. Having the same problem twice may mean some hidden issue

File: RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h [hunk 10]
```diff
@@ -4,32 +4,40 @@
...
   OuterHitPhiPrediction( 
       const Range & phiAtVertex, 
       const Range & curvature, 
-      float originRBound, 
-      const Margin & tolerance = Margin(0,0))
+      float originRBound) 
     : thePhiAtVertex(phiAtVertex), theCurvature(curvature),
-      theOriginRBound (originRBound), theTolerance(tolerance) { } 
+      theOriginRBound (originRBound) {
+       assert(theCurvature.max()>0);
```
Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: Hmm, in our TSG tests we suddenly hit this assert:
```
# Conditions read from  CMS_CONDITIONS  via FrontierProd 
%MSG-i ThreadSetup:  (NoModuleName) 07-Jun-2016 19:17:49 CEST pre-events
setting # threads 4
%MSG
07-Jun-2016 19:18:00 CEST  Initiating request to open file file:RelVal_Raw_GRun_DATA.root
07-Jun-2016 19:18:01 CEST  Successfully opened file file:RelVal_Raw_GRun_DATA.root
isBH_ 18:13:18:0
Begin processing the 1st record. Run 272762, Event 41669009, LumiSection 51 at 07-Jun-2016 19:18:58.423 CEST
Begin processing the 2nd record. Run 272762, Event 41468301, LumiSection 51 at 07-Jun-2016 19:18:58.425 CEST
Begin processing the 3rd record. Run 272762, Event 41643921, LumiSection 51 at 07-Jun-2016 19:18:58.438 CEST
Begin processing the 4th record. Run 272762, Event 41555597, LumiSection 51 at 07-Jun-2016 19:18:58.459 CEST
cmsRun: /build/cmsbuild/jenkins-workarea/workspace/build-any-ib/w/tmp/BUILDROOT/6eb14a8c47f34962dae9df674c18b15e/opt/cmssw/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-05-0000/src/RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h:23: OuterHitPhiPrediction::OuterHitPhiPrediction(const Range&, const Range&, float): Assertion `theCurvature.max()>0' failed.
```
This is obviously problematic!

Comment (VinInn) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: would like to know how to reproduce

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: For example, using: CMSSW_8_1_X_2016-06-08-2300, add PR #14811,
and execute ```addOnTests.py -t hlt_data_GRun``` - this works.
Then edit the file ```RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h:23```
and put back the ```assert``` and run the above addOn again. Then there is a crash!

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: BTW, it seems related to the GT used: if I use the HLT GT ```80X_dataRun2_HLT_frozen_v12```
it crashes. If I use the reco GT ```80X_dataRun2_v14``` it does not crash. But of course we
need to use the HLT GT, so is there something missing in there?

Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @Martin-Grunewald if a record would be missing it would crash before the first event.
I would argue that the difference of calibrations between `80X_dataRun2_v14` (data re-reco, no IOVs from 2016), vs `80X_dataRun2_HLT_frozen_v12` (online GT, IOVs from 2016).

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: Crash also disappears on using 80X_dataRun2_HLT_relval_v11, but I thought
80X_dataRun2_HLT_frozen_v12 is always mor up to date than 80X_dataRun2_HLT_relval_v11
?? (See separate e-mail)

Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: `80X_dataRun2_HLT_frozen_v12` is indeed more updated than `80X_dataRun2_HLT_relval_v11`, exactly as `80X_dataRun2_HLT_frozen_v12` is more updated in `80X_dataRun2_v14`. The issue probably lies in fresh conditions....

Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @Martin-Grunewald I beg your pardon, actually `80X_dataRun2_HLT_relval_v11` has later snapshot time than `80X_dataRun2_HLT_frozen_v12` (because it was created later in time):

https://cms-conddb.cern.ch/cmsDbBrowser/diff/Prod/gts/80X_dataRun2_HLT_frozen_v12/80X_dataRun2_HLT_relval_v11

even if on the surface only the L1 menu tag is different, you can see in the table called `IOV differences` that many (actually 51) tags have developed new IOVs that are not seen in `80X_dataRun2_HLT_frozen_v12`  because they were inserted after its snapshot time.
Can you try with the "open" GT that goes directly online, I mean: `80X_dataRun2_HLT_v12`  ?

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: ```80X_dataRun2_HLT_v12``` works also!
(BTW, note that in all these tests I add the L1T menu explicitly via toGet)

Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @Martin-Grunewald I just realized that you are trying to use a fixed snapshot time Global Tag on data taken in the relative future with respect to the snapshot time itself.
The GT `80X_dataRun2_HLT_frozen_v12` has snapshot set to:  2016-04-20 06:59:54 UTC.
The run you are using was taken on: 2016.05.06 22:27:33 UTC.
This simply means that you are seeing plain wrong conditions for that IOV!
As an example, the last IOV for  `RunInfoRcd` (that drives the magnetic field map you load) seen in the GT `80X_dataRun2_HLT_frozen_v12` is at run 270395, 

https://cmswbm.web.cern.ch/cmswbm/cmsdb/servlet/RunSummary?RUN=270395&SUBMIT=Submit

which happens to be a 0T run !
And indeed if you dump the magnet currents you'll get:
```
$ conddb --noLimit list runinfo_start_31X_hlt | grep 270395
270395       2016-04-20 06:32:39  786aaad4b7bcfab74d0d10fde752e22b637168bf  RunInfo      
$ conddb dump 786aaad4b7bcfab74d0d10fde752e22b637168bf | grep current
	<m-start-current>1.063721371e+01</m-start-current>
	<m-stop-current>1.063721371e+01</m-stop-current>
	<m-avg-current>1.063721371e+01</m-avg-current>
	<m-max-current>1.063721371e+01</m-max-current>
	<m-min-current>1.063721371e+01</m-min-current>
```
while it should be:
```
$ conddb --noLimit list runinfo_start_31X_hlt | grep 272762
272762       2016-05-06 22:27:38  bcef1447160d669c62328b485ce1cfdd4c4a4efd  RunInfo      
$ conddb dump bcef1447160d669c62328b485ce1cfdd4c4a4efd > dump_bcef1447160d669c62328b485ce1cfdd4c4a4efd.xml | grep current
	<m-start-current>1.814441211e+04</m-start-current>
	<m-stop-current>1.814441211e+04</m-stop-current>
	<m-avg-current>1.814441211e+04</m-avg-current>
	<m-max-current>1.814441211e+04</m-max-current>
	<m-min-current>1.814441211e+04</m-min-current>
```
Reconstructing a 3.8T run with 0T conditions is bound to fail, IMHO.
Frozen GTs are meant for relvals and there is **NO** guarantee at all they are going to cover anything else outside run ranges covered in relvals. If you wish to test an arbitrary run please use the open snapshot GTs made available here:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideFrontierConditions#HLT_reconstruction_Global_Tags

otherwise wait for 2016 data relvals to become available.

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: Hi Marco, these open snapshot GTs such as 80X_dataRun2_HLT_v12 have no symbolic entry in autoCond.py,
could you make one? Alternatively, could you make a new version of the 80X_dataRun2_HLT_frozen_v GT
with a snapshot time covering up to this weeks TS?
And if 80X_dataRun2_HLT_frozen_v is for relvals, why is there a GT called 80X_dataRun2_HLT_relval_v?

**Comment by VinInn** [comment]
Reconstructing a 3.8T run with 0T conditions  explain the failing assert.
Actually this points out that the whole code in  RectangularEtaPhiTrackingRegion is not protected against 0T field.
 In any case removing the assert is correct if we run tracking at HLT for 0T and we want to continue to reproduce the current behaviour.

File: RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h (hunk 10 above)
Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: let's clear your confusion then :) 
1) I definitely don't want to introduce open snapshot GTs (such as 80X_dataRun2_HLT_v*), because this (if used by mistake in any production like workflow) will result in non-reproducibilities.
2) The PR you are referring to is not yet merged
3) RelVals should definitely use the `auto:run2_hlt_relval`as they are correctly doing.
4) the one and only reason why I haven't cleaned up yet the key `'auto:run2_hlt` from autoCond is because TSG is using it all over the place (https://github.com/cms-sw/cmssw/search?utf8=%E2%9C%93&q=run2_hlt).
5) producing a new GT (actually same GT, but with newer snapshot time)  for the `80X_dataRun2_HLT_frozen_v`can be done (we are in any case planning to make a PR to add new records in 81x soon).

In summary:
   * using `auto:run2_hlt_relval` instead of `auto:run2_hlt` is up to you, depending if you need the L1 menu from O2O or a fixed version (the latest) available in the `auto:run2_hlt_relval` key.
   * the fact that in this particular case you've got wrong conditions in one case in bad conditions in another is just an accident due to the particular time of the snapshot. In both case is still **wrong** to reconstuct data taken past the snapshot

Comment (Martin-Grunewald) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @mmusich 
Thanks a lot for the explanation.
Following this, I shall change from ```auto:run2_hlt``` to ```auto:run2_hlt_relval``` for the TSG/HLTrigger side (and I assume DQM should do the same given the use cases shown in your link), so that we can then get rid of ```auto:run2_hlt``` in autoCond.py.

Comment (fwyzard) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: > In both case is still wrong to reconstuct data taken past the snapshot 

I agree, but is there some kind of protection or check to detect when this happens ?

Comment (mmusich) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @fwyzard not that I am aware of, might be useful to throw some warning. @ggovi

Comment (ggovi) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: @fwyzard @Martin-Grunewald
To further clarify: specify a snapshot in a GT make sense only when the underlying condition set is 'a good one' and is frozen. In practice, the GT with snapshot are targeted to specific (non-evolving) data sets. The usage for data 'in the future' wrt the snapshot is in general inappropriate.
That said, we could add a warning to remark the usage outside the snapshot boundary, although this doesn't necessarily consists in something wrong in itself...
It's more about to set up a policy and how strict we want to be ( in the usage of GT with snapshot ). We might even decide to 'close the iovs' and generate an error.

Comment (Dr15Jones) [inline RecoTracker/TkTrackingRegions/src/OuterHitPhiPrediction.h]: Closed IOVs are the mechanism the framework expects to be used to catch such problems.
